### ISSUE TITLE
design a flexible way to pass various locations

### ISSUE THREAD

**Issue by sopos** [beakerlib/beakerlib#51]
as a library maintainer I want to be able to have my library in a custom location

**Comment by sopos** [comment]
A variable BEAKERLIB_LIBRARIES will be used to define custom locations.

current state:
the library is defined by a component and a library name, it is written in format `<component>/<library name>` and it must be located in a directory `<component>/Library/<library name>` 

`BEAKERLIB_LIBRARIES` is a space-separated list of following references:

I will be using library `my-comp/basic-1` as an illustrative example

1. reference to a common place for libraries:
URL(s) to clone `<URL>/<LIBRARY>, LIBRARY here cannot contain slash (/) and other special characters so they are replaced by dashes (-), thus the actual library repo is `<component>-<library name>`
e.g. URL `https://github.com/beakerlib-libraries` and library would be actually located in `https://github.com/beakerlib-libraries/my-comp-basic-1`

1. reference of a single library
LIBRARY:URL
e.g. `my-comp/basic-1:https://github.com/beakerlib-libraries/my-comp-basic-1`

**Comment by sopos** [comment]
While I was experimenting with `git` I found out that if I try to clone non-existing repo `git` asks for credentials. This needs to be somehow worked around as it is quite expected scenario.

**Comment by sopos** [comment]
> A variable BEAKERLIB_LIBRARIES will be used to define custom locations.
> 
> current state:
> the library is defined by a component and a library name, it is written in format `<component>/<library name>` and it must be located in a directory `<component>/Library/<library name>`
> 
> `BEAKERLIB_LIBRARIES` is a space-separated list of following references:
> 
> I will be using library `my-comp/basic-1` as an illustrative example
> 
>     1. reference to a common place for library repos:
>        URL(s) to clone `<URL>/<LIBRARY>, LIBRARY here cannot contain slash (/) and other special characters so they are replaced by dashes (-), thus the actual library repo is `-`e.g. URL`[https://github.com/beakerlib-libraries`](https://github.com/beakerlib-libraries%60) and library would be actually located in `https://github.com/beakerlib-libraries/my-comp-basic-1`
> 
>     2. reference of a single library in dedicated repo
>        (LIBRARY)URL
>        e.g. `(my-comp/basic-1)https://github.com/beakerlib-libraries/my-comp-basic-1`
> 
>     3. reference of a library inside of a general repo
>        (LIBRARY)URL
>        e.g. `(my-comp/basic-1)https://github.com/some/repo/raw/master/libraries/my-comp-basic-1.yaml`

Actually I have probably better and more generic way to describe the reference now:
Actually only one common patter for describing all three reference variants.
```
(library regexp pattern)URL
```
a special keyword `LIBRARY` in the reference would be substituted by the actual LIBRARY (with dashes)
 
e.g.
1. reference to a common place for library repos
```
(.*)https://github.com/beakerlib-libraries/LIBRARY/raw/master/lib.yaml
```
1. reference of a single library in dedicated repo
```
(my-comp/basic-1)https://github.com/beakerlib-libraries/my-comp-basic-1/raw/master/lib.yaml
```
1. reference of a library inside of a general repo
```
(my-comp/basic-1)https://github.com/some/repo/raw/master/libraries/my-comp-basic-1.yaml
```
1. reference of more libraries inside of a general repo
```
(my-comp/basic-1|my-comp/basic-2)https://github.com/some/repo/raw/master/libraries/LIBRARY.yaml
```

**Comment by sopos** [comment]
Actually only one common pattern to describe all three reference variants and more.
```
(library regexp pattern)URL
```
a special keyword `LIBRARY` in the reference would be substituted by the actual LIBRARY (with dashes)
 
e.g.
1. reference to a common place for library repos
```
(.*)https://github.com/beakerlib-libraries/LIBRARY/raw/master/lib.yaml
```
1. reference of a single library in dedicated repo
```
(my-comp/basic-1)https://github.com/beakerlib-libraries/my-comp-basic-1/raw/master/lib.yaml
```
1. reference of a library inside of a general repo
```
(my-comp/basic-1)https://github.com/some/repo/raw/master/libraries/my-comp-basic-1.yaml
```
1. reference of more libraries inside of a general repo
```
(my-comp/basic-1|my-comp/basic-2)https://github.com/some/repo/raw/master/libraries/LIBRARY.yaml
```

Basically when looking for a library the URL would be considered in the LIBRARY matches the regexp pattern, otherwise it would be just wasting of time.

The yaml file should contain an information how to get whole library. Currently I'm thinking of just a link to the archive and links to raw files, but it might be extended in the future by other methods.
The advantage of this solution is that there is no need for further dependencies for beakerlib.

**Comment by sopos** [comment]
To allow to dive the library down into the component's repo it would make sense to define `COMPONENT` and `NAME` which would act as `LIBRARY=COMPONENT/NAME`.

**Comment by sopos** [comment]
format of the definition yaml file:
```
url:
base_url:
files:
```
`url` - reference to a single file (complete url)
`base_url` - a URL to be used together with `files`
`files` - array of items to be added to base_url to construct the final URL for a download

example 1:
```
url: https://github.com/some/repo/raw/master/lib.sh
```
example 2:
```
base_url: https://github.com/some/repo/raw/master/
files:
  - lib.sh
  - Makefile
  - additional_files/x
  - additional_files/y
```
note that `Makefile` might be necessary in case of other libraries required

**Comment by sopos** [comment]
another section `requires` and/or `` requires_lib` might be added to somehow express the library requirements

**Comment by sopos** [comment]
actually, this is not needed anymore as it is prepared by the schedulers, like tmt
